main/src/ca/uwaterloo/flix/language/phase/jvm/GenChannelClasses.scala

```diff
@@ -0,0 +1,705 @@
+
+    // Return the value
+    mv.visitLabel(labelEndHandler)
+    if (channelType != JvmType.Object) {
+      mv.visitMethodInsn(INVOKEVIRTUAL, channelType.getBoxedTypeString, channelType.getUnboxingMethod, AsmOps.getMethodDescriptor(Nil, channelType), false)
+    }
+    mv.visitInsn(iRet)
+    mv.visitMaxs(4, 4)
+    mv.visitEnd()
+  }
+
+  /**
+    * Generates the `putValue()` method of the `classType` with value of type `channelType`
+    *
+    * public `classType` putValue(tpe value) throws InterruptedException {
+    *     this.lock.lock();
+    *
+    *     try {
+    *         while(this.isFull()) {
+    *             this.awaitGetters();
+    *         }
+    *
+    *         this.offer(value);
+    *         this.signalGetters();
+    *     } finally {
+    *         this.lock.unlock();
+    *     }
+    *
+    *     return this;
+    * }
+    */
+  def genPutValue(classType: JvmType.Reference, channelType: JvmType, visitor: ClassWriter)(implicit root: Root, flix: Flix): Unit = {
+    val iLoad = AsmOps.getLoadInstruction(channelType)
+    val mv = visitor.visitMethod(ACC_PUBLIC, "putValue", AsmOps.getMethodDescriptor(List(channelType), classType), null, Array(JvmName.InterruptedException.toInternalName))
+    val labelStart = new Label()
+    val labelEnd = new Label()
+    val labelHandler = new Label()
+    val loopStart = new Label()
+    val loopEnd = new Label()
+    val labelReturn = new Label()
+    mv.visitCode()
+
+    // Lock
+    mv.visitVarInsn(ALOAD, 0)
+    mv.visitFieldInsn(GETFIELD, classType.name.toInternalName, "lock", JvmType.Lock.toDescriptor)
+    mv.visitMethodInsn(INVOKEINTERFACE, JvmType.Lock.name.toInternalName, "lock", AsmOps.getMethodDescriptor(Nil, JvmType.Void), true)
+
+    mv.visitTryCatchBlock(labelStart, labelEnd, labelHandler, null)
+
+    // Try
+    mv.visitLabel(labelStart)
+
+    // Loop
+    mv.visitLabel(loopStart)
+    mv.visitVarInsn(ALOAD, 0)
+    mv.visitMethodInsn(INVOKEVIRTUAL, classType.name.toInternalName, "isFull", AsmOps.getMethodDescriptor(Nil, JvmType.PrimBool), false)
+    mv.visitJumpInsn(IFEQ, loopEnd)
+    mv.visitVarInsn(ALOAD, 0)
+    mv.visitMethodInsn(INVOKEVIRTUAL, classType.name.toInternalName, "awaitGetters", AsmOps.getMethodDescriptor(Nil, JvmType.Void), false)
+    mv.visitJumpInsn(GOTO, loopStart)
+    mv.visitLabel(loopEnd)
+
+    // Offer
+    mv.visitVarInsn(ALOAD, 0)
+    mv.visitVarInsn(iLoad, 1)
+    mv.visitMethodInsn(INVOKEVIRTUAL, classType.name.toInternalName, "offer", AsmOps.getMethodDescriptor(List(channelType), JvmType.PrimBool), false)
+
+    // TODO: Use the variable instead of just popping it
+    mv.visitInsn(POP)
+
+    // Signal All
+    mv.visitVarInsn(ALOAD, 0)
+    mv.visitMethodInsn(INVOKEVIRTUAL, classType.name.toInternalName, "signalGetters", AsmOps.getMethodDescriptor(Nil, JvmType.Void), false)
+
+    // TODO: Clear Selects
+    //mv.visitVarInsn(ALOAD, 0)
```

Comment:
What is happening here?

If selects are not cleared that could cause problems (and definitely memory leaks).